File: backend/app/schemas/save.py
from typing import Any
from datetime import datetime

from pydantic import BaseModel, Field

from app.core.versions import SUPPORTED_CONTRACT_VERSION
from app.schemas.game import GameError, GameState


class SaveRecord(BaseModel):
    save_id: str
    name: str
    created_at: datetime
    updated_at: datetime
    save_format_version: int
    state: GameState
    meta: dict[str, Any] = Field(default_factory=dict)


class SaveSummary(BaseModel):
    save_id: str
    name: str
    created_at: datetime
    updated_at: datetime
    save_format_version: int


class SaveCreateRequest(BaseModel):
    contract_version: int
    request_id: str
    name: str
    state: GameState
    meta: dict[str, Any] = Field(default_factory=dict)


class SaveCreateResponse(BaseModel):
    ok: bool
    contract_version: int
    request_id: str
    save: SaveRecord | None
    error: GameError | None

    @classmethod
    def unsupported_version(cls, request_id: str, contract_version: int) -> "SaveCreateResponse":
        return cls(
            ok=False,
            contract_version=SUPPORTED_CONTRACT_VERSION,
            request_id=request_id,
            save=None,
            error=GameError(
                code="UNSUPPORTED_VERSION",
                message=f"Unsupported contract version: {contract_version}",
                details={"contract_version": contract_version},
            ),
        )


class SaveListResponse(BaseModel):
    ok: bool
    contract_version: int
    request_id: str | None
    saves: list[SaveSummary]
    error: GameError | None


class SaveGetSingleResponse(BaseModel):
    ok: bool
    contract_version: int
    request_id: str | None
    save: SaveRecord | None
    error: GameError | None


class SaveUpdateRequest(BaseModel):
    contract_version: int
    request_id: str
    name: str
    state: GameState
    meta: dict[str, Any] = Field(default_factory=dict)


class SaveUpdateResponse(BaseModel):
    ok: bool
    contract_version: int
    request_id: str
    save: SaveRecord | None
    error: GameError | None
    
    @classmethod
    def unsupported_version(cls, request_id: str, contract_version: int) -> "SaveUpdateResponse":
        return cls(
            ok=False,
            contract_version=SUPPORTED_CONTRACT_VERSION,
            request_id=request_id,
            save=None,
            error=GameError(
                code="UNSUPPROTED_VERSION",
                message=f"Unsupported contract version: {contract_version}",
                details={"contract_version": contract_version},
            )
        )


class SaveDeleteResponse(BaseModel):
    ok: bool
    contract_version: int
    request_id: str | None
    deleted: bool
    save_id: str
    error: GameError | None

File: frontend/src/types/save.ts
import type { GameError, GameState } from "./game";

export interface SaveRecord {
  save_id: string;
  name: string;
  created_at: string;
  updated_at: string;
  save_format_version: number;
  state: GameState;
  meta: Record<string, unknown>;
}

export interface SaveSummary {
  save_id: string;
  name: string;
  created_at: string;
  updated_at: string;
  save_format_version: number;
}

export interface SaveCreateRequest {
  contract_version: number;
  request_id: string;
  name: string;
  state: GameState;
  meta: Record<string, unknown>;
}

export interface SaveUpdateRequest {
  contract_version: number;
  request_id: string;
  name: string;
  state: GameState;
  meta: Record<string, unknown>;
}

export interface SaveCreateResponse {
  ok: boolean;
  contract_version: number;
  request_id: string;
  save: SaveRecord | null;
  error: GameError | null;
}

export interface SaveListResponse {
  ok: boolean;
  contract_version: number;
  request_id: string | null;
  saves: SaveSummary[];
  error: GameError | null;
}

export interface SaveGetSingleResponse {
  ok: boolean;
  contract_version: number;
  request_id: string | null;
  save: SaveRecord | null;
  error: GameError | null;
}

export interface SaveUpdateResponse {
  ok: boolean;
  contract_version: number;
  request_id: string;
  save: SaveRecord | null;
  error: GameError | null;
}

export interface SaveDeleteResponse {
  ok: boolean;
  contract_version: number;
  request_id: string | null;
  deleted: boolean;
  save_id: string;
  error: GameError | null;
}
